src/mpid/ch4/src/ch4r_rma.h

```diff
@@ -243,175 +243,213 @@ static inline int MPIDI_CH4I_do_get(void *origin_addr,
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_CH4U_RGET);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_CH4U_RGET);
+    am_hdr.req_ptr = (uint64_t) sreq;
+    am_hdr.origin_count = origin_count;
+    am_hdr.do_get = 0;  /* FIXME: separate packet type for get_acc. */
+
+    if (HANDLE_GET_KIND(origin_datatype) == HANDLE_KIND_BUILTIN) {
+        am_hdr.origin_datatype = origin_datatype;
+    }
+    else {
+        am_hdr.origin_datatype = (dt_ptr) ? dt_ptr->basic_type : MPI_DATATYPE_NULL;
+        MPID_Datatype_get_size_macro(am_hdr.origin_datatype, basic_type_size);
+        am_hdr.origin_count = (basic_type_size > 0) ? data_sz / basic_type_size : 0;
+    }
 
-    mpi_errno = MPIDI_CH4I_do_get(origin_addr, origin_count, origin_datatype,
-                                  target_rank, target_disp, target_count,
-                                  target_datatype, win, request);
+    am_hdr.target_count = target_count;
+    am_hdr.target_datatype = target_datatype;
+    am_hdr.target_disp = target_disp;
+    am_hdr.op = op;
+    am_hdr.win_id = MPIDI_CH4U_WIN(win, win_id);
+    am_hdr.src_rank = win->comm_ptr->rank;
+
+    /* MPIDI_CS_ENTER(); */
+    OPA_incr_int(&MPIDI_CH4U_WIN(win, outstanding_ops));
+    /* MPIDI_CS_EXIT(); */
+
+    MPIDI_CH4U_REQUEST(sreq, req->areq.data_sz) = data_sz;
+    if (HANDLE_GET_KIND(target_datatype) == HANDLE_KIND_BUILTIN) {
+        am_hdr.n_iov = 0;
+        MPIDI_CH4U_REQUEST(sreq, req->areq.dt_iov) = NULL;
+
+        mpi_errno = MPIDI_NM_am_isend(target_rank, win->comm_ptr, MPIDI_CH4U_ACC_REQ,
+                                      &am_hdr, sizeof(am_hdr), origin_addr,
+                                      (op == MPI_NO_OP) ? 0 : origin_count,
+                                      origin_datatype, sreq, NULL);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+        goto fn_exit;
+    }
+
+    MPIDI_Datatype_get_size_dt_ptr(target_count, target_datatype, data_sz, dt_ptr);
+    am_hdr.target_datatype = dt_ptr->basic_type;
+    am_hdr.target_count = dt_ptr->n_builtin_elements;
+
+    segment_ptr = MPIDU_Segment_alloc();
+    MPIR_Assert(segment_ptr);
+
+
+    MPIDU_Segment_init(NULL, target_count, target_datatype, segment_ptr, 0);
+    last = data_sz;
+    MPIDU_Segment_count_contig_blocks(segment_ptr, 0, &last, &num_iov);
+    n_iov = (int) num_iov;
+    MPIR_Assert(n_iov > 0);
+    am_hdr.n_iov = n_iov;
+    dt_iov = (struct iovec *) MPL_malloc(n_iov * sizeof(struct iovec));
+    MPIR_Assert(dt_iov);
+
+    last = data_sz;
+    MPIDU_Segment_pack_vector(segment_ptr, 0, &last, dt_iov, &n_iov);
+    MPIR_Assert(last == (MPI_Aint) data_sz);
+    MPL_free(segment_ptr);
+
+    am_iov[0].iov_base = &am_hdr;
+    am_iov[0].iov_len = sizeof(am_hdr);
+    am_iov[1].iov_base = dt_iov;
+    am_iov[1].iov_len = sizeof(struct iovec) * am_hdr.n_iov;
+    MPIDI_CH4U_REQUEST(sreq, req->areq.dt_iov) = dt_iov;
+
+    if ((am_iov[0].iov_len + am_iov[1].iov_len) <= MPIDI_NM_am_hdr_max_sz()) {
```

Comment:
We had decided to get rid of this entire hdr_max_size API.  Maybe add a FIXME here and get back to it later?